### PR TITLE
Feature/xamarin essentials events

### PR DESCRIPTION
**Pull request by michaelstonis**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Feature

**What is the current behavior? (You can also link to an open issue here)**
There are no event bindings for Xamarin.Essentials today

**What is the new behavior (if this is a feature change)?**
This adds the ability to provide event binding for Xamarin.Essentials

**What might this PR break?**
Not likely anything, but there were some changes to the event binding program.

**Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:
I would potentially consider this a _first pass_. Everything in the Essentials world is a static class, so binding the same way we did before is not possible. Right now, the extensions are added to `object`, so as soon as you do a `using Xamarin.Essentials;`, every object is going to have these extension methods available. I don't love this, but I really couldn't think of a better approach.

### CODE REVIEW

File: src/EventBuilder/EventBuilder.csproj [hunk 1]
```diff
@@ -18,9 +18,14 @@
     <PackageReference Include="CommandLineParser" Version="1.9.71" />
     <PackageReference Include="Microsoft.Web.Xdt" Version="2.1.1" />
     <PackageReference Include="Mono.Cecil" Version="0.9.6.1" />
-    <PackageReference Include="NuGet.Core" Version="2.10.1" />
     <PackageReference Include="Nustache" Version="1.15.3.7" />
     <PackageReference Include="Polly" Version="3.0.0" />
     <PackageReference Include="Serilog" Version="1.5.14" />
+    <PackageReference Include="NuGet.Core" Version="2.14.0" />
   </ItemGroup>
+<ItemGroup>
```
Comment (olevett) [inline src/EventBuilder/EventBuilder.csproj]: This formatting seems a bit inconsistent with the others, e.g. indent off by one

File: src/EventBuilder/Platforms/Essentials.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+                Log.Debug("Using Xamarin Essentials {Version} released on {Published}", package.Version, package.Published);
+                Log.Debug("{ReleaseNotes}", package.ReleaseNotes);
+            });
+
+            var xamarinForms =
+                Directory.GetFiles(packageUnzipPath,
+                    "Xamarin.Essentials.dll", SearchOption.AllDirectories);
+
+            var latestVersion = xamarinForms.Last();//x => x.Contains("netstandard2.0"));
```
Comment (olevett) [inline src/EventBuilder/Platforms/Essentials.cs]: is the x.contains needed?

File: src/EventBuilder/Cecil/StaticEventTemplateInformation.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            {
+                "Windows.UI.Xaml.Data.PropertyChangedEventHandler",
+                "global::System.ComponentModel.PropertyChangedEventHandler"
+            },
+            {"Windows.Foundation.EventHandler", "EventHandler"},
+            {"Windows.Foundation.EventHandler`1", "EventHandler"},
+            {"Windows.Foundation.EventHandler`2", "EventHandler"}
+        };
+
+        private static string RenameBogusWinRTTypes(string typeName)
```
Comment (olevett) [inline src/EventBuilder/Cecil/StaticEventTemplateInformation.cs]: Given this is only being used for Xamarin.Essentials, do we need this slightly odd bodge?

File: src/EventBuilder/Cecil/StaticEventTemplateInformation.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            var publicTypesWithEvents = targetAssemblies
+                .SelectMany(x => SafeTypes.GetSafeTypes(x))
+                .Where(x => x.IsPublic && !x.HasGenericParameters)
+                .Select(x => new {Type = x, Events = GetPublicEvents(x)})
+                .Where(x => x.Events.Length > 0)
+                .ToArray();
+
+            var garbageNamespaceList = new[]
+            {
+                "Windows.UI.Xaml.Data",
```
Comment (olevett) [inline src/EventBuilder/Cecil/StaticEventTemplateInformation.cs]: Likewise, are these namespaces sensible to exclude?

File: src/EventBuilder/Cecil/StaticEventTemplateInformation.cs (hunk 3 above)
Comment (michaelstonis) [reply src/EventBuilder/Cecil/StaticEventTemplateInformation.cs]: I am going to kick it all out for now.

File: src/EventBuilder/Cecil/StaticEventTemplateInformation.cs (hunk 4 above)
Comment (michaelstonis) [reply src/EventBuilder/Cecil/StaticEventTemplateInformation.cs]: Same, I am cleaning all of the, seemingly, non-relevant stuff out of here.

File: src/EventBuilder/Platforms/Essentials.cs (hunk 2 above)
Comment (michaelstonis) [reply src/EventBuilder/Platforms/Essentials.cs]: Originally I had it pinned at `netstandard2.0`, but then had to switch it to try to just get the last one. I am going to pin it at `netstandard2.0`.

**Comment by olevett** [comment]
Thanks @michaelstonis!

@ghuntley not sure who has the ability to push this out as an extra nuget package, is this something you could help with?

**Comment by Mrnikbobjeff** [comment]
I would suggest tabling this until [this](https://github.com/xamarin/Essentials/pull/375) has been delt with in their repository. I reckon this would break existing code.

**Comment by michaelstonis** [comment]
I will keep an eye on the Essentials package and update this once they finalize their event structure.

**Comment by Mrnikbobjeff** [comment]
@michaelstonis The event structure has been finalized. That block has been cleared.

**Comment by michaelstonis** [comment]
As soon as they release a new nuget with these in them, I will get this updated.

**Comment by glennawatson** [comment]
They released a new version a couple days ago https://www.nuget.org/packages/Xamarin.Essentials

**Comment by michaelstonis** [comment]
Cool. I am on it.

File: src/EventBuilder/Platforms/Essentials.cs [hunk 5]
```diff
@@ -46,7 +46,7 @@ public Essentials()
                 Directory.GetFiles(packageUnzipPath,
                     "Xamarin.Essentials.dll", SearchOption.AllDirectories);
 
-            var latestVersion = xamarinForms.First(x => x.Contains("netstandard2.0"));
+            var latestVersion = xamarinForms.First(x => x.Contains("netstandard1.0"));
```
Comment (glennawatson) [inline src/EventBuilder/Platforms/Essentials.cs]: Wonder if it's worthwhile not putting the version in?

File: src/EventBuilder/CommandLineOptions.cs [hunk 6]
```diff
@@ -17,7 +17,8 @@ public enum AutoPlatform
         WPF,
         XamForms,
         UWP,
-        Winforms
+        Winforms,
```
Comment (glennawatson) [inline src/EventBuilder/CommandLineOptions.cs]: This is going to be fun doing TVOS who wins first :)

File: src/EventBuilder/EventBuilder.csproj [hunk 7]
```diff
@@ -1,26 +1,30 @@
...
     <PackageReference Include="Microsoft.Web.Xdt" Version="2.1.1" />
     <PackageReference Include="Mono.Cecil" Version="0.9.6.1" />
-    <PackageReference Include="NuGet.Core" Version="2.10.1" />
     <PackageReference Include="Nustache" Version="1.15.3.7" />
     <PackageReference Include="Polly" Version="3.0.0" />
     <PackageReference Include="Serilog" Version="1.5.14" />
+    <PackageReference Include="NuGet.Core" Version="2.14.0" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Update="XamarinEssentialsTemplate.mustache">
```
Comment (glennawatson) [inline src/EventBuilder/EventBuilder.csproj]: This one is None, other mustache above is Content, can we keep them the same type and put them into the same ItemGroup?

**Comment by glennawatson** [comment]
BTW @michaelstonis I think you just need to do a conflict resolve on the files at the moment to get this ready for finishing the PR, and very minor minor comments in review. Then I think we are ready to pull.

**Comment by glennawatson** [comment]
Hey @olevett you mentioned we might need a new NuGet package for the events? Should I get Oren/Geoffrey to help out?

File: src/EventBuilder/EventBuilder.csproj (hunk 7 above)
Comment (michaelstonis) [reply src/EventBuilder/EventBuilder.csproj]: can do!

**Comment by glennawatson** [comment]
Thanks @michaelstonis

**Pull request merged**